### PR TITLE
feat(standups): rich text in slack notifications

### LINKED ISSUE #8651: TipTap x Slack: Format rich tiptap content for Slack
TipTap standup responses posted to slack currently just use the basic `plaintextContent`, which strips links, styling, etc.

We should support special formatting for all of the tiptap extensions we're using

AC:
- [ ] Slack formatting support for (in rough priority order)
  - [ ] Links
  - [ ] Bullets
  - [ ] Numbers
  - [ ] Code blocks
  - [ ] Bold
  - [ ] Italics
  - [ ] Strikethrough
- [ ] (optional) special slack formatting support for
  - [ ] Mentions

### PR DESCRIPTION
**Pull request by BartoszJarocki**
# Description

Fixes #8651

## Demo

Screenshot 2023-08-23 at 16 41 56

## Testing scenarios

Testing scenarios assume Slack is already integrated

- [ ] Open standup meeting, write an update that contains styling, close the meeting, see the slack notification

## Final checklist

- [x] I checked the [code review guidelines](../docs/codeReview.md)
- [x] I have added [Metrics Representative](../docs/codeReview.md#metrics-representative) as reviewer(s) if my PR invovles metrics/data/analytics related changes
- [x] I have performed a self-review of my code, the same way I'd do it for any other team member
- [x] I have tested all cases I listed in the testing scenarios and I haven't found any issues or regressions
- [x] Whenever I took a non-obvious choice I added a comment explaining why I did it this way
- [x] I added the label `One Review Required` if the PR introduces only minor changes, does not contain any architectural changes or does not introduce any new patterns and I think one review is sufficient'
- [x] PR title is human readable and could be used in changelog

### CODE REVIEW

**Comment by BartoszJarocki** [comment]
@jmtaber129 @mattkrick there are at least two possible solutions

First of all, tiptap does not support providing output as markdown, which is what slack supports. that leaves us two options

1. Quick, just like the example implementation in this PR - ie. add a [turndown library that converts HTML to markdown]([mixmark-io/turndown: 🛏 An HTML to Markdown converter written in JavaScript](https://github.com/mixmark-io/turndown)) and just use it

Pros:
- quick, and in theory configurable with plugins (ie. we could still implement mentions etc)

Cons
- for example html headers won't be displayed in slack `section` component that we use to format notifications, so it looks like that
Screenshot 2023-08-23 at 09 12 03

we could still override the conversion made by this library and just make the headers bold or something like that.
 
2. Do the conversion ourselves, this gives us more flexibility 

Pros
- we can split the message into multiple slack blocks and use not only markdown section, but also headers etc. 
- more flexibility 

Cons
- more time consuming 

any thoughts? If the priority is to ship fast, we can try to tweak he library, if we can spend a little bit more time, we can provide our own implementation that'd be more future proof.

**Comment by BartoszJarocki** [comment]
decode do go with custom impl, I don't want to spend any time fixing external library quirks 🤷‍♂️

**Comment by BartoszJarocki** [comment]
@jmtaber129 do you mind taking a look?

File: packages/server/utils/tiptap/convertToMarkdown.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import {JSONContent} from '@tiptap/core'
+
+const LIST_SPACING = '   '
+
+const tiptapNodeHandlersLookup: {[type: string]: (node: JSONContent, depth?: number) => string} = {
```
Comment (mattkrick) [inline packages/server/utils/tiptap/convertToMarkdown.tsx:5]: +2 50 LOCs & no external deps? pretty sweet!

**Comment by mattkrick** [comment]
I absolutely love your solution, smaller bespoke packages always get my vote.
curious, did you check out https://github.com/aguingand/tiptap-markdown? it's sooo much larger than your solution, so i was wondering what extra functionality it might give?

**Review by jmtaber129 — CHANGES_REQUESTED** [review]
Code looks good, and really nice approach! Just a few unsupported nodes+marks things:

-1 Mentions aren't supported - If I mention a user, nothing appears where their name is
-1 Inline code snippets like `these` appear as plaintext
+1 The horizontal rule (`<hr/>`) nodes aren't included - this isn't too important (we didn't _really_ support these until last week) and there's not a super elegant way to handle this without switching this approach to pure slack blocks (or refactor the recursive solution to an iterative one) AFAIK, but we could insert a line like `--------` just to show that the response had some form of horizontal divider.
+1 Quote blocks - our CSS basically just indents with no other styling when a user inserts a quote block (i.e. what you get with a `>` in slack), but it would be nice to include support for this in slack anyway.

**Comment by BartoszJarocki** [comment]
@jmtaber129 thanks for pointing these out! 

1. mentions - I've added support to just display it like a normal text - here's an example 
Screenshot 2023-08-25 at 13 37 28
to make it a real mention, we'd need to map Parabol user id into Slack user id. at this point I'm not sure of the effort needed to to that, but I'd suggest merging it as is and handling this as a separate issue (as this is a big improvement over existing functionality already). 

2. inline code - added!

3. horizonal rule, how can I add that to our standup response so I can test? 😅 

4. black quotes - added!

**Comment by BartoszJarocki** [comment]
@jmtaber129 added support for horizontal lines too. it'll be just `---` as Slack does not seem to support it. 
Screenshot 2023-08-25 at 16 46 58

can you take a look again?

**Review by jmtaber129 — APPROVED** [review]
One small change needed to fix mentions, otherwise LGTM

File: packages/server/utils/tiptap/convertToMarkdown.tsx [hunk 2]
```diff
@@ -33,23 +54,41 @@ const tiptapNodeHandlersLookup: {[type: string]: (node: JSONContent, depth?: num
...
     return `\`\`\`${markdownContent}\`\`\`` // Triple backticks to denote a code block in Slack
-  }
+  },
+  blockquote: ({content = []}) => `> ${content.map((node) => convertToMarkdown(node)).join('')}`,
+  mention: ({attrs}) => {
+    const {
+      /** id - TODO: here our internal Parabol user id would need to bo mapped to Slack user id */
+      label
+    } = attrs!
+    return `<@${label}>`
```
[the remark this answers is not in the document]
Comment (BartoszJarocki) [reply packages/server/utils/tiptap/convertToMarkdown.tsx]: fixed. it required a little bit of investigation as it turns out Slack tries to [automatically parse mentions, links etc. ](https://api.slack.com/reference/surfaces/formatting#automatic-parsing__disabling-automatic-parsing)

**Comment by BartoszJarocki** [comment]
@mattkrick do you mind taking a final look? thanks!

**Pull request merged**